### PR TITLE
NIP -50

### PR DESCRIPTION
**Pull request by srivastavaAbhi**
Header is verified in CDR files.

@frankhuster : Please review it.
   Also, suggest me that while sorting a file,is it right to remove header from sorted file?
   I have not removed it yet and processed it as you coded.

### CODE REVIEW

**Comment by frankhuster** [comment]
If you make sure the file is sorted in descending order then the first line
will still be the header,
so change this line
<https://github.com/motech-implementations/mim/blob/master/imi/src/main/java/org/motechproject/nms/imi/service/impl/SortHelper.java#L68>
to

String command = String.format("%s -t , -k 1,1r -k 4,4n -o %s %s",
getSortBinary(),

should do the trick

ps: note I added 'r' to the -k 1,1 argument


On Mon, Jul 6, 2015 at 4:44 PM, srivastavaAbhi <notifications@github.com>
wrote:

> Header is verified in CDR files.
>
> @frankhuster <https://github.com/frankhuster> : Please review it.
> Also, suggest me that while sorting a file,is it right to remove header
> from sorted file?
> I have not removed it yet and processed it as you coded.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/motech-implementations/mim/pull/457
> Commit Summary
>
>    - NIP -50
>
> File Changes
>
>    - *M*
>    imi/src/main/java/org/motechproject/nms/imi/service/impl/CdrFileServiceImpl.java
>    <https://github.com/motech-implementations/mim/pull/457/files#diff-0>
>    (229)
>    - *M*
>    testing/src/test/java/org/motechproject/nms/testing/it/utils/CdrHelper.java
>    <https://github.com/motech-implementations/mim/pull/457/files#diff-1>
>    (10)
>
> Patch Links:
>
>    - https://github.com/motech-implementations/mim/pull/457.patch
>    - https://github.com/motech-implementations/mim/pull/457.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/motech-implementations/mim/pull/457>.
>



-- 
Frank Huster
Mobile Health Innovations
Grameen Foundation
+1.206.325.6690 x201

**Comment by srivastavaAbhi** [comment]
@frankhuster  : I have done the correction. Please check

File: imi/src/main/java/org/motechproject/nms/imi/service/impl/CsvHelper.java [hunk 1]
```diff
@@ -141,4 +141,31 @@ public static CallDetailRecordDto csvLineToCdr(String line) {
...
+
+        String[] fields = line.split(",");
+
+        if (fields.length != FieldName.FIELD_COUNT.ordinal()) {
+            throw new IllegalArgumentException(String.format(
+                    "Invalid field count, expecting %d but received %d", FieldName.FIELD_COUNT.ordinal(),
+                    fields.length));
+        }
+        if (!(header.equalsIgnoreCase(line))) {
+            throw new IllegalArgumentException(String.format("Header in the file is not correct"));
```
Comment (frankhuster) [inline imi/src/main/java/org/motechproject/nms/imi/service/impl/CsvHelper.java]: No need for `String.format()`

File: imi/src/main/java/org/motechproject/nms/imi/service/impl/CsvHelper.java [hunk 2]
```diff
@@ -141,4 +141,31 @@ public static CallDetailRecordDto csvLineToCdr(String line) {
...
+     */
+    public static void validateCdrHeader(String line) {
+
+        String header = "RequestId,Msisdn,CallId,AttemptNo,CallStartTime,CallAnswerTime,CallEndTime," +
+                "CallDurationInPulse,CallStatus,LanguageLocationId,ContentFile,MsgPlayStartTime,MsgPlayEndTime," +
+                "CircleId,OperatorId,Priority,CallDisconnectReason,WeekId";
+
+        String[] fields = line.split(",");
+
+        if (fields.length != FieldName.FIELD_COUNT.ordinal()) {
```
Comment (frankhuster) [inline imi/src/main/java/org/motechproject/nms/imi/service/impl/CsvHelper.java]: Not needed, the `equalsIgnoreCase()` call below is sufficient

File: testing/src/test/java/org/motechproject/nms/testing/it/utils/CdrHelper.java [hunk 3]
```diff
@@ -384,6 +387,13 @@ private File doMakeCdrFile(String dir, int numInvalidLines) throws IOException {
         return file;
     }
 
+    public static String writeHeader() {
```
Comment (frankhuster) [inline testing/src/test/java/org/motechproject/nms/testing/it/utils/CdrHelper.java]: that doesn't write anything, should be called header() or, even better, make that a string constant.

**Comment by frankhuster** [comment]
@srivastavaAbhi is traveling, I'll try to take that over tomorrow and make the changes I suggested in the CR.

**Pull request merged**